src/GitHub/App.hs

```diff
@@ -0,0 +1,152 @@
+baseURL = "api.github.com"
+
+type BaseURL = Text 
+type InstallationId = Text 
+
+
+-- | Github installation Auth token 
+data GTAToken = GTAToken 
+    {
+        instToken  :: !Text,      -- ^ Installation token 
+        expireAt   :: !POSIXTime  -- ^ expiration time 
+    }
+
+-- | Main installation auth info for application. 
+-- it containes immutable app information and app installation token
+-- which shoud be updated at least once per hour  
+data InstallationAuth = InstallationAuth 
+    {
+        _appId          :: !Int,           -- ^ application id 
+        _appPrivateKey  :: !PrivateKey,    -- ^ Private key to sign token requests  
+        _installationId :: !Text,          -- ^ Installation id 
+        _token          :: IORef GTAToken  -- ^ Installation Auth token. 
+    } 
+
+makeLenses '' InstallationAuth
+
+instance FromJSON GTAToken where
+    parseJSON = withObject "Github installation token" $ \o -> do 
+        instToken   <- o .: "token"  
+        expireAtM   <- expPOSIXtime <$> o .: "expires_at"
+        case expireAtM of 
+            Nothing       -> fail "Failed to parse expiraion time."
+            Just expireAt -> return GTAToken{..}
+
+-- | newtype for failed JSON Web token generation.
+newtype FailedGenerateJWT = FailedGenerateJWT Text
+
+instance Show.Show FailedGenerateJWT where
+    show (FailedGenerateJWT err) = toString err
+
+instance Exception FailedGenerateJWT
+
+-- | Checks if current app installation token is not expired. If it is not, than it just returns token 
+-- with OAuth wrapping. 
+-- Otherwise it call function that gets new token from github and writes it to IORef, and also returns that new token. 
+authenticateInstallation :: InstallationAuth -> IO Auth
+authenticateInstallation instAuth = do 
+    gtaToken    <- readIORef (instAuth ^. token)
+    currentTime <- getPOSIXTime
+    if (expireAt gtaToken) - currentTime >= bufferTime 
+    then return $ OAuth $ encodeUtf8 $ instToken gtaToken
+    else do 
+        let applicationId = instAuth ^. appId
+        OAuth <$> encodeUtf8 <$> instToken <$> (join $ readIORef <$> renewInstAuthToken instAuth) 
+
+
+-- | Gets new token from github and writes it to IORef in given InstallationAuth. Also returns that token. 
+renewInstAuthToken :: InstallationAuth -> IO (IORef GTAToken) 
+renewInstAuthToken instAuth = do 
+    time    <- getPOSIXTime
+    let tkn = instAuth ^. token
+    case (makeJWT time (instAuth ^. appId) (instAuth ^. appPrivateKey)) of 
+        Left  err  -> throwM (FailedGenerateJWT err) 
+        Right jwt  -> do 
+            t <- request (https baseURL /: "installations" /: (instAuth ^. installationId) /: "access_tokens") mempty jwt    
+            writeIORef tkn t 
+            return tkn
+
+-- | Creates new InstallationAuth value. Useing this function is the only way to create InstallationAuth value, 
+-- because we don't export constructors. 
+createInstAuth :: Int -> PrivateKey -> InstallationId -> IO InstallationAuth 
```

Comment:
Isn’t it the same function as above?

Reply:
Yes. I made 2 almost the same functions, because i can't create empty IORef, now i use MVar.